src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobTask.cs

```diff
@@ -126,96 +124,207 @@ public async Task ExecuteAsync(ReindexJobRecord reindexJobRecord, WeakETag weakE
                     _reindexJobRecord.Count = countOnlyResults.TotalCount.Value;
 
                     // Query first batch of resources
-                    await ProcessQueryAsync(queryStatus, cancellationToken);
+                    await ProcessQueryAsync(queryStatus, jobSemaphore, cancellationToken);
                 }
-                else
+
+                var queryTasks = new List<Task<ReindexJobQueryStatus>>();
+                var queryCancellationTokens = new Dictionary<ReindexJobQueryStatus, CancellationTokenSource>();
+
+                // while not all queries are finished
+                while (_reindexJobRecord.QueryList.Where(q =>
+                    q.Status == OperationStatus.Queued ||
+                    q.Status == OperationStatus.Running).Any())
                 {
-                    // check to see if queries are queued
-                    // TODO: this while loop is temporary until we multithread this task so multiple threads can
-                    // processes queries
-                    while (_reindexJobRecord.QueryList.Where(q => q.Status == OperationStatus.Queued).Any())
+                    if (_reindexJobRecord.QueryList.Where(q => q.Status == OperationStatus.Queued).Any())
                     {
                         // grab the next query from the list which is labeled as queued and run it
                         var query = _reindexJobRecord.QueryList.Where(q => q.Status == OperationStatus.Queued).OrderBy(q => q.LastModified).FirstOrDefault();
+                        CancellationTokenSource queryTokensSource = new CancellationTokenSource();
+                        queryCancellationTokens.Add(query, queryTokensSource);
+
+#pragma warning disable CS4014 // Suppressed as we want to continue execution and begin processing the next query while this continues to run
+                        queryTasks.Add(ProcessQueryAsync(query, jobSemaphore, queryTokensSource.Token));
+#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
 
-                        await ProcessQueryAsync(query, cancellationToken);
+                        _logger.LogInformation($"Reindex job task using {queryTasks.Count} threads");
```

Comment:
This is not entirely true, if too many tasks are created they can be awaiting activation, before running